File: ios/App/Sources/Features/Artifacts/ArtifactTitleGenerator.swift
import Foundation

/// Names artifacts via **Lightweight Tasks** (Apple Intelligence or a
/// user-linked model). Falls back to the first useful content line.
enum ArtifactTitleGenerator {
    static let maxTitleLength = 64

    static var isOnDeviceAvailable: Bool {
        LightweightTaskRunner.appleFoundationAvailable
    }

    static func suggestTitle(for content: String) async -> String {
        let heuristic = heuristicTitle(from: content)
        let sample = clip(content, limit: 2_400)
        guard !sample.isEmpty else { return heuristic }

        let system = """
        You name saved chat artifacts (documents, code, reports) for a coding assistant app.
        Reply with ONLY a short title: 3 to 8 words.
        No quotes, no trailing punctuation, no emoji, no explanation, no markdown.
        Capture what the artifact is (e.g. "Instagram audio research notes", "Login form SwiftUI").
        Prefer concrete nouns over vague words like "document" or "response".
        """

        let user = """
        Artifact content:
        \(sample)

        Title:
        """

        if let generated = await LightweightTaskRunner.complete(
            system: system,
            user: user,
            maxTokens: 28
        ), let clean = sanitize(generated) {
            return clean
        }

        return heuristic
    }

    static func heuristicTitle(from content: String) -> String {
        let fence = content.range(of: #"```(\w+)?"#, options: .regularExpression)
        if let fence {
            // `lineRange(for:)` is non-optional — do not bind with `if let`.
            let lineRange = content.lineRange(for: fence.lowerBound..<fence.lowerBound)
            let fenceLine = content[lineRange].trimmingCharacters(in: .whitespacesAndNewlines)
            if fenceLine.hasPrefix("```") {
                let lang = fenceLine.dropFirst(3).trimmingCharacters(in: .whitespacesAndNewlines)
                let after = content[lineRange.upperBound...]
                for raw in after.split(separator: "\n", omittingEmptySubsequences: true).prefix(3) {
                    let line = raw.trimmingCharacters(in: .whitespaces)
                    if line.hasPrefix("```") { break }
                    if !line.isEmpty {
                        let snippet = String(line.prefix(40))
                        if lang.isEmpty { return sanitizeHeuristic(snippet) }
                        return sanitizeHeuristic("\(lang): \(snippet)")
                    }
                }
                if !lang.isEmpty { return sanitizeHeuristic("\(lang) code") }
            }
        }

        for raw in content.split(separator: "\n", omittingEmptySubsequences: true) {
            var line = raw.trimmingCharacters(in: .whitespaces)
            if line.hasPrefix("```") { continue }
            if line.hasPrefix("#") {
                line = String(line.drop(while: { $0 == "#" })).trimmingCharacters(in: .whitespaces)
            }
            if !line.isEmpty {
                return sanitizeHeuristic(String(line.prefix(maxTitleLength)))
            }
        }
        return "Artifact"
    }

    static func sanitize(_ raw: String) -> String? {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if text.lowercased().hasPrefix("<think>") {
            if let end = text.range(of: "</think>", options: .caseInsensitive) {
                text = String(text[end.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        guard !text.isEmpty else { return nil }

        if let newline = text.firstIndex(of: "\n") {
            text = String(text[..<newline]).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        let wrappers: [Character] = ["\"", "'", "`", "*", "“", "”", "‘", "’"]
        while let first = text.first, let last = text.last,
              wrappers.contains(first), wrappers.contains(last),
              text.count > 2 {
            text = String(text.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        for prefix in ["Title:", "Name:", "Artifact:"] {
            if text.lowercased().hasPrefix(prefix.lowercased()) {
                text = String(text.dropFirst(prefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }

        if text.hasSuffix("."), text.count <= 48, !text.dropLast().contains(".") {
            text = String(text.dropLast())
        }

        guard !text.isEmpty else { return nil }
        if text.count <= maxTitleLength { return text }
        return String(text.prefix(maxTitleLength - 1)) + "…"
    }

    private static func sanitizeHeuristic(_ text: String) -> String {
        sanitize(text) ?? "Artifact"
    }

    private static func clip(_ text: String, limit: Int) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.count <= limit { return trimmed }
        return String(trimmed.prefix(limit)) + "…"
    }
}

File: android/app/src/main/kotlin/app/roamsocket/android/ui/artifacts/ArtifactTitleGenerator.kt
package app.roamsocket.android.ui.artifacts

/**
 * Generates a short display title for an artifact.
 *
 * iOS prefers **Lightweight Tasks** (Apple Foundation Model or a
 * user-linked model) and falls back to a heuristic. Android currently
 * has no on-device model, so we always use the heuristic — the public
 * surface mirrors the iOS API so the call site in `ChatViewModel`
 * stays the same and we can plug in a real backend later
 * (the LightweightTasks port will provide one).
 *
 * Ports `ArtifactTitleGenerator` from
 * `ios/App/Sources/Features/Artifacts/ArtifactTitleGenerator.swift`.
 */
object ArtifactTitleGenerator {
    const val MAX_TITLE_LENGTH: Int = 64

    /**
     * Suggest a title. Currently always returns the heuristic. When
     * Android gets an on-device model (LightweightTasks port), this
     * will try that first and fall back to the heuristic.
     */
    suspend fun suggestTitle(content: String): String = heuristicTitle(content)

    /**
     * First useful line: prefer the language tag from a code fence, or
     * a heading, or the first non-empty, non-fence line.
     */
    fun heuristicTitle(content: String): String {
        val fenceMatch = Regex("""```(\w+)?""").find(content)
        if (fenceMatch != null) {
            val fenceStart = fenceMatch.range.first
            val lineStart = content.lastIndexOf('\n', startIndex = fenceStart).let { it + 1 }
            val lineEnd = content.indexOf('\n', startIndex = fenceStart).let { if (it < 0) content.length else it }
            val fenceLine = content.substring(lineStart, lineEnd).trim()
            if (fenceLine.startsWith("```")) {
                val lang = fenceLine.removePrefix("```").trim()
                val after = content.substring(lineEnd + 1)
                for (raw in after.split('\n').take(3)) {
                    val line = raw.trim()
                    if (line.isEmpty()) continue
                    if (line.startsWith("```")) break
                    val snippet = line.take(40)
                    return sanitize(if (lang.isEmpty()) snippet else "$lang: $snippet")
                }
                if (lang.isNotEmpty()) return sanitize("$lang code")
            }
        }

        for (raw in content.split('\n')) {
            var line = raw.trim()
            if (line.startsWith("```")) continue
            if (line.startsWith("#")) {
                line = line.dropWhile { it == '#' }.trim()
            }
            if (line.isNotEmpty()) {
                return sanitize(line.take(MAX_TITLE_LENGTH))
            }
        }
        return "Artifact"
    }

    /**
     * Strip common wrappers and prefixes an LLM might add
     * (quotes, leading "Title:", trailing period, …).
     */
    fun sanitize(raw: String): String {
        var text = raw.trim()
        if (text.isEmpty()) return "Artifact"

        // Drop everything after the first newline (model only returns the title).
        val newline = text.indexOf('\n')
        if (newline >= 0) {
            text = text.substring(0, newline).trim()
        }

        // Strip matched quote-style wrappers ("…", '…", `…`).
        val wrappers = charArrayOf('"', '\'', '`', '*', '“', '”', '‘', '’')
        while (
            text.length > 2 &&
            wrappers.contains(text.first()) &&
            wrappers.contains(text.last())
        ) {
            text = text.substring(1, text.length - 1).trim()
        }

        // Strip leading "Title:" / "Name:" / "Artifact:".
        for (prefix in listOf("Title:", "Name:", "Artifact:")) {
            if (text.startsWith(prefix, ignoreCase = true)) {
                text = text.substring(prefix.length).trim()
            }
        }

        // Drop a single trailing period on short titles.
        if (text.endsWith('.') && text.length <= 48 && !text.dropLast(1).contains('.')) {
            text = text.dropLast(1)
        }

        if (text.isEmpty()) return "Artifact"
        if (text.length <= MAX_TITLE_LENGTH) return text
        return text.take(MAX_TITLE_LENGTH - 1) + "…"
    }
}
